rules/composer/src/ValueObject/ComposerModifier/AddPackageToRequireDev.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Composer\ValueObject\ComposerModifier;
+
+use Rector\Composer\Modifier\ComposerModifier;
+use Rector\Composer\Contract\ComposerModifier\ComposerModifierInterface;
+use Rector\Composer\ValueObject\Version\Version;
+
+/**
+ * Only adds package to require-dev section, if package is already in composer data, nothing happen
+ * @see \Rector\Composer\Tests\ValueObject\ComposerModifier\AddPackageToRequireDevTest
+ */
+final class AddPackageToRequireDev implements ComposerModifierInterface
+{
+    /** @var string */
+    private $packageName;
+
+    /** @var Version */
+    private $version;
+
+    /**
+     * @param string $packageName name of package (vendor/package)
+     * @param string $version target package version (1.2.3, ^1.2, ~1.2.3 etc.)
+     */
+    public function __construct(string $packageName, string $version)
+    {
+        $this->packageName = $packageName;
+        $this->version = new Version($version);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function modify(array $composerData): array
+    {
+        if (!isset($composerData[ComposerModifier::SECTION_REQUIRE][$this->packageName]) && !isset($composerData[ComposerModifier::SECTION_REQUIRE_DEV][$this->packageName])) {
+            $composerData[ComposerModifier::SECTION_REQUIRE_DEV][$this->packageName] = $this->version->getVersion();
+        }
+
+        return $composerData;
+    }
```

Comment:
I'm not sure if this is value object or a service.
It modified outer data `$composerData` without keeping a state, so it's a service.

It should be separated to value object and modifier service

Reply:
this is some older version, actual is:

```
public function modify(ComposerJson $composerJson): ComposerJson
    {
        $composerJson->addRequiredDevPackage($this->packageName, $this->version->getVersion());
        return $composerJson;
    }
```

Reply:
I see. That's the same, just with different argument.

Reply:
So basically, you want me, to split all these classes to two classes:
1) ValueObject with only constructor (and probably some getters)
and
2) service with modify method which will have 2 arguments ComposerJson $composerJson and ValueObject for getting configuration (package name, version etc.).

Right?

Reply:
@TomasVotruba Hello, I didn't get an answer here :) I just want to say that first implementation get this comment from you: `The code looks very clean and is easy to read` :) With 2x more classes it imho will not be clean and easy to read. Also first implementation was really easy to add another modifiers like ChangeName, ChangeDescription etc. because you just manipulate an array in modify() method of ComposerModifierInterface implementation. It is also possible now with ComposerJson object (if it has all needed methods), but adding another layer just unnecessarily add some complexity which is imho not needed here... But it is your package, so if you want, I will rewrite it as you wish, just answer to my question. As I am thinking now, maybe I will have to add some Factory to select correct service depending on which value object will be processing

Reply:
Sorry, I've missed it. Yes you got it right 👍